### PR TITLE
Use generic delete dialog for resetting, cleanup

### PR DESCRIPTION
**Pull request by maxr96**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~Server: I added multiple integration tests (Spring) related to the features~
- [x] ~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~
- [x] ~Server: I implemented the changes with a good performance and prevented too many database calls~
- [x] ~Server: I documented the Java code using JavaDoc style.~
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

This should resolve the issue #874. Archiving will be replaced by Downloading Repos functionality both on Course Administration and View pages for Programming Exercise.

### Description

We have a lot of code duplication in creating the dialog, as well as separate popup URLs for each dialog. This PBI is focused on resolving this issues. Additionally, the dialog should wait until the delete request is executed and not close if the error is thrown.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Select Reset button on Programming Exercise (1).
4. Select Cleanup button on Programming Exercise (2).
5. Select Reset button on Quiz Exercise (3).
6. Select Delete button on Programming Exercise.
7. Click 'Delete' in the dialog, it should take a bit of time until the request to server is sent and we get the result in the client (4).

#### Reset Programming Exercise (Course Administration)
Screenshot_2019-11-23 Courses
---
#### Cleanup Programming Exercise (Course Administration and Programming View pages)
Screenshot_2019-10-14 Courses(1)
---
#### Reset Quiz Exercise (Course Administration)
Screenshot_2019-10-14 Courses(2)
---

### CODE REVIEW

File: src/main/webapp/app/entities/text-exercise/text-exercise.component.ts [hunk 1]
```diff
@@ -9,20 +9,22 @@ import { ActivatedRoute } from '@angular/router';
...
     @Input() textExercises: TextExercise[];
+    closeDialogTrigger: boolean;
 
     constructor(
         private textExerciseService: TextExerciseService,
         private courseExerciseService: CourseExerciseService,
         courseService: CourseService,
         translateService: TranslateService,
-        private jhiAlertService: JhiAlertService,
+        protected jhiAlertService: JhiAlertService,
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/webapp/app/entities/text-exercise/text-exercise.component.ts]: Right, thanks for checking

File: src/main/webapp/app/shared/delete-dialog/delete-dialog.model.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    entityTitle: string;
+
+    // i18n key, that will be translated
+    deleteQuestion: string;
+
+    // i18n key, if undefined no safety check will take place (input name of the entity)
+    deleteConfirmationText?: string;
+
+    // object with check name as a key and i18n key as a value, check names will be used for the return statement
+    additionalChecks?: { [key: string]: string };
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/webapp/app/shared/delete-dialog/delete-dialog.model.ts]: Additional checks is passed into the dialog to show check boxes with corresponding translation strings in the dialog. They are available for example when deleting programming exercise

**Comment by maxr96** [comment]
> Hi @maxr96 I just added a few minor comments here. I could not test it on the test server yet because the deployment failed. Maybe you could look into the checks here in this PR in the meantime? I will test it thoroughly once I can deploy this branch again.

Hi @lovaiible . Thanks for checking my pull request and sorry for the not working building. After merging the latest changes from develop everything passes again.

**Comment by maxr96** [comment]
> Hi @maxr96 I have tested your implementation and it looks good to me.
> 
> However there is one change that I would request:
> The UI of the programming exercise buttons has changed and the `manage hints`, `download repos` and `cleanup` moved to the programming exercise detail page, see screenshot
> image
> 
> Could you remove those from the `programming-exercise` component?
> image

You are right @lovaiible, I didn't check that change. Now it should be fine.
Screenshot_2019-10-25 Courses

File: src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.service.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@ import { SERVER_API_URL } from 'app/app.constants';
 import { FileUploadExercise } from './file-upload-exercise.model';
 import { createRequestOption } from 'app/shared';
 import { ExerciseService } from 'app/entities/exercise';
+import { throwError } from 'rxjs';
```
Comment (lovaiible) [inline src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.service.ts]: You can remove the unused import :)

File: src/main/webapp/app/entities/programming-exercise/programming-exercise.component.ts [hunk 4]
```diff
@@ -9,23 +9,29 @@ import { ActivatedRoute, Router } from '@angular/router';
...
 
     constructor(
         private programmingExerciseService: ProgrammingExerciseService,
         private courseExerciseService: CourseExerciseService,
+        private exerciseService: ExerciseService,
         private accountService: AccountService,
         private jhiAlertService: JhiAlertService,
-        private modalService: NgbModal,
         private router: Router,
+        private modalService: NgbModal,
```
Comment (lovaiible) [inline src/main/webapp/app/entities/programming-exercise/programming-exercise.component.ts]: We don't use any modals anymore, right?

File: src/main/webapp/i18n/en/instructorDashboard.json [hunk 5]
```diff
@@ -17,7 +17,7 @@
         "students": "Students",
         "cleanup": {
             "title": "Cleanup",
-            "question": "Are you sure you want to delete ALL student build plans for the programming exercise {{ exerciseTitle }} in the course {{courseTitle}}? This action can NOT be undone!"
+            "question": "Are you sure you want to delete ALL student build plans for the programming exercise {{ title }}? This action can NOT be undone!"
```
Comment (lovaiible) [inline src/main/webapp/i18n/en/instructorDashboard.json]:
```suggestion
            "question": "Are you sure you want to delete ALL student build plans for the programming exercise <strong>{{ title }}</strong>? This action can NOT be undone!"
```

File: src/main/webapp/i18n/de/instructorDashboard.json [hunk 6]
```diff
@@ -17,7 +17,7 @@
         "students": "Studenten",
         "cleanup": {
             "title": "Aufräumen",
-            "question": "Sollen wirklich ALLE Studierenden Build Pläne für die Programmieraufgabe {{ exerciseTitle }} im Kurs {{ courseTitle }} dauerhaft gelöscht werden? Diese Aktion kann NICHT rückgängig gemacht werden!"
+            "question": "Sollen wirklich ALLE Studierenden Build Pläne für die Programmieraufgabe {{ title }} dauerhaft gelöscht werden? Diese Aktion kann NICHT rückgängig gemacht werden!"
```
Comment (lovaiible) [inline src/main/webapp/i18n/de/instructorDashboard.json]:
```suggestion
            "question": "Sollen wirklich ALLE Studierenden Build Pläne für die Programmieraufgabe <strong>{{ title }}</strong> dauerhaft gelöscht werden? Diese Aktion kann NICHT rückgängig gemacht werden!"
```

**Comment by lovaiible** [comment]
The `download repositories` button is in the scores page actually (I'm sorry if I mistook it in my earlier comment). Instead of this button, the `Archive` button is missing.
image

**Comment by maxr96** [comment]
> The `download repositories` button is in the scores page actually (I'm sorry if I mistook it in my earlier comment). Instead of this button, the `Archive` button is missing.
> image

Thank you @lovaiible for checking my PR again. As discussed previously with @krusche we decided that `Archive` button should be removed completely and replaced with `Download Repository` button. Since we have already `Download Repository` in the Scores page, maybe we can remove it from here, but I am not sure about this.

File: src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
 
 @Directive({ selector: '[jhiDeleteButton]' })
 export class DeleteButtonDirective implements OnInit {
     @Input() entityTitle: string;
     @Input() deleteQuestion: string;
     @Input() deleteConfirmationText: string;
     @Input() additionalChecks?: { [key: string]: string };
+    @Input() actionType: ActionType = ActionType.Delete;
     @Output() delete = new EventEmitter<{ [key: string]: boolean }>();
+    @Input() set closeDialogTrigger(value: boolean) {
```
Comment (ungaralex) [inline src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts]: Why aren't we using some kind of observable here?

**Comment by krusche** [comment]
@maxr96 please resolve the merge conflicts. 
Also double check the popups. For the reset programming exercise case, you miss one option as far as I can tell from the screenshot. Please make sure that this refactoring does not change the functionality of the buttons!!!

File: src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts (hunk 7 above)
Comment (maxr96) [reply src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts]: I like that idea, but in my opinion it would be an overkill for each component to create an observable, that will only emit once. Passing `boolean` value is a lot simpler in terms of reusability.

**Comment by maxr96** [comment]
> @maxr96 please resolve the merge conflicts.
> Also double check the popups. For the reset programming exercise case, you miss one option as far as I can tell from the screenshot. Please make sure that this refactoring does not change the functionality of the buttons!!!

@krusche I resolved the conflicts.
The functionality stays the same. In case of reset dialog for programming exercise I indeed removed this option, because it was not used in the code. Ticking the checkbox doesn't change the behavior of the reset operation. Reset endpoint that is called also doesn't have additional options that can be passed.

**Comment by krusche** [comment]
Then the problem might be the missing option in the REST Call.

**Comment by maxr96** [comment]
> Then the problem might be the missing option in the REST Call.

As far as I understand resetting programming exercise should at least delete all participations and results. So the checkbox can then say `Additionally delete all student repositories and student build plans ... `. The same as we have in the delete dialog. 
Current behavior of delete dialog is to delete all participations, results, student repositories and build plans. Should we split it or leave it as it is? @krusche

**Comment by krusche** [comment]
please resolve the merge conflicts and make sure to follow the new angular design guidelines described in #1081

**Comment by krusche** [comment]
@maxr96 what is the status of this PR? Seems to be outdated and has a lot of conflicting files now. Should we close this for now and reopen it when more progress is made?

**Comment by maxr96** [comment]
> @maxr96 what is the status of this PR? Seems to be outdated and has a lot of conflicting files now. Should we close this for now and reopen it when more progress is made?

I merged the latest changes, most of which where related to the bug fix made by @ungaralex . This PR is still ready fir review.

**Comment by krusche** [comment]
> > @maxr96 please resolve the merge conflicts.
> > Also double check the popups. For the reset programming exercise case, you miss one option as far as I can tell from the screenshot. Please make sure that this refactoring does not change the functionality of the buttons!!!
> 
> @krusche I resolved the conflicts.
> The functionality stays the same. In case of the reset dialog for programming exercise I indeed removed this option, because it was not used in the code. Ticking the checkbox doesn't change the behavior of the reset operation. Furthermore, reset endpoint that is called doesn't have an additional option that can be passed.

I just checked the reset dialog and the user has to check the checkbox, otherwise the user cannot click on Delete. So this is more a confirmation that the user knows what Reset actually means and does not accidentally perform it.

image

image

This means we definitely need this option, so please reactivate it!

By the way: I had a request some long time ago, that we only close the delete dialog if the server response came in. What happens with this? Did you finally implement it? I think it should be part of this PR, because this is important for users that the dialog does not close immediately, but only when the server call was successful!!!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 8]
```diff
@@ -289,7 +289,7 @@ private StatsForInstructorDashboardDTO populateCommonStatistics(Exercise exercis
             return forbidden();
         exerciseService.cleanup(id, deleteRepositories);
         log.info("Cleanup build plans was successful for Exercise : {}", id);
-        return ResponseEntity.ok().headers(HeaderUtil.createAlert(applicationName, "Cleanup was successful. Repositories have been deleted: " + deleteRepositories, "")).build();
+        return ResponseEntity.ok().build();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java]: why did you change this and remove the feedback?

Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java]: Because I show the alerts now on the client after I received successful response.

**Comment by maxr96** [comment]
> > > @maxr96 please resolve the merge conflicts.
> > > Also double check the popups. For the reset programming exercise case, you miss one option as far as I can tell from the screenshot. Please make sure that this refactoring does not change the functionality of the buttons!!!
> > 
> > 
> > @krusche I resolved the conflicts.
> > The functionality stays the same. In case of the reset dialog for programming exercise I indeed removed this option, because it was not used in the code. Ticking the checkbox doesn't change the behavior of the reset operation. Furthermore, reset endpoint that is called doesn't have an additional option that can be passed.
> 
> I just checked the reset dialog and the user has to check the checkbox, otherwise the user cannot click on Delete. So this is more a confirmation that the user knows what Reset actually means and does not accidentally perform it.
> 
> image
> 
> image
> 
> This means we definitely need this option, so please reactivate it!
> 
> By the way: I had a request some long time ago, that we only close the delete dialog if the server response came in. What happens with this? Did you finally implement it? I think it should be part of this PR, because this is important for users that the dialog does not close immediately, but only when the server call was successful!!!

I updated the reset dialog for programming in a similar manner we have it for resetting quiz exercise
image
Yes I implemented the logic for dialog to wait for the server response and it works for all delete, cleanup and reset dialogs that use the `deleteDialogButton` directive. I also mentioned it in the description of this PR.

**Comment by beatrisbu** [comment]
I tested the changes that you described and everything seems to work for me. I could not reproduce the problem which Anh mentioned.

**Comment by maxr96** [comment]
> @maxr96 Good job refactoring the delete dialogues and reducing the duplicated code!
> 
> I have found an issue before this PR can be merged:
> When deleting programming exercises, the delete button is not disabled and does not show a loading icon, compared to other buttons in Artemis (e.g. the generate programming exercise button). The user can still click on `delete` multiple times (since the modal is open until the exercise is deleted) and trigger an error:
> image

Thank you for checking my PR @lovaiible . Indeed I was able to reproduce the issue you mentioned. To fix it I took suggestion from @ungaralex to use Observable. Now if the delete action was triggered the delete button is blocked until we get any result from the action. If the result is an error, the button is re-enabled, if the delete action succeeded dialog will be closed.

File: src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts [hunk 9]
```diff
@@ -48,14 +52,22 @@ export class DeleteButtonDirective implements OnInit {
...
             additionalChecks: this.additionalChecks,
+            actionType: this.actionType,
+            delete: this.delete,
+            dialogError: this.dialogError,
         };
-        this.deleteDialogService.openDeleteDialog(deleteDialogData).subscribe((additionalChecksValues: { [key: string]: boolean }) => {
-            this.delete.emit(additionalChecksValues);
-        });
+        this.deleteDialogService.openDeleteDialog(deleteDialogData).subscribe(
+            () => {},
```
Comment (madwau) [inline src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts]: Why subscribe if nothing is done?

Comment (maxr96) [reply src/main/webapp/app/shared/delete-dialog/delete-button.directive.ts]: You are right here in the previous implementation we were emitting delete event here. Now this is not used, so I moved the subscribe to the service itself. There we need to clear the modalRef when the dialog is closed

**Comment by madwau** [comment]
> > I executed the test steps and everything works as expected.
> > Some minor UI flaws I noticed:
> > ```
> > 1. When a button is disabled (such as the "Reset" button when the exercise name hasn't been typed in yet), prefer the cursor `not-allowed` over `pointer` on hover to communicate that the button can't be clicked in its current state.
> >    OnPaste 20191203-200504
> > 
> > 2. Alerts appear in two places, once in the dialog and once in the background page (ignore the error itself, I provoked it by setting throttling to offline in the Chrome dev tools).
> >    OnPaste 20191203-201140
> > ```
> 
> Thank you for the review @madwau
> 
> 1. It is a good point to make it more clear to the user that the button can't be clicked. Implemented it
>    Screenshot_20191204_001855
> 2. I agree that maybe it is a bit overwhelming, but this is handled by `jhiAlertService` and I don't see a way to change it. Basically, if you specify `jhi-alert` in the template it will show the error in all the rendered templates that have this tag. We need this tag in both templates in this case.

1. Thanks!
2. Ok, if there is no easy way to resolve this, let's accept it for now and maybe address it in a future PR focusing on just that.

**Comment by maxr96** [comment]
> Looks like a test is now failing since the return value of `openDeleteDialog` changed with your last commit.

Thank you for checking. Strange that I didn't receive any notification on my email, since that usually happens when the PR fails.

**Comment by krusche** [comment]
Good findings, @HanyaElhashemy 
Especially the naming and bold needs to be consistent for the reset dialog as well. 
The clicking outside is not an important issue. The problem here is that the div of the text continues so we cannot really prevent this easily.

@maxr96 please fix the issues mentioned by @HanyaElhashemy

**Comment by beatrisbu** [comment]
Tested the changes on the test server and the they work.
The problems mentioned by Hanya and the first problem mentioned by Martin have been addressed successfully. The exercises' names are bold everywhere. When I click outdide of the checkbox, it does not get unchecked. The reset button is clearly disabled everywhere, before the name of the exercise has been typed.
After the TypeScript issue has also been adressed, I can give my approval.

**Comment by maxr96** [comment]
> Great Job! I only have some small points I would like to address:
> 
>     * Also when I click outside the checkbox it gets unchecked
>       image

So apparently the problem for you was not updating the browser after uploading new build version on the server. For me the dialog looks like this. 
Screenshot_2019-12-04 Courses
Regarding the clicking outside I agree with @krusche that it is not a big issue and would take some time to fix. Especially, that we had this behavior all the time

**Comment by HanyaElhashemy** [comment]
@krusche @beatrisbu  I now see that these issues were addressed after testing again, @maxr96 thank you for that. I edited my previous comment accordingly ^^
The unchecking issue is the only thing I can still reproduce but like you, all said it is not crucial so I also approve this PR. Thanks for clarifying!

**Comment by beatrisbu** [comment]
> > Great Job! I only have some small points I would like to address:
> > ```
> > * Also when I click outside the checkbox it gets unchecked
> >   image
> > ```
> 
> So apparently the problem for you was not updating the browser after uploading new build version on the server. For me the dialog looks like this.
> Screenshot_2019-12-04 Courses
> Regarding the clicking outside I agree with @krusche that it is not a big issue and would take some time to fix. Especially, that we had this behavior all the time

Not eon the matter - I do not experience that issue

**Comment by HanyaElhashemy** [comment]
> > > Great Job! I only have some small points I would like to address:
> > > ```
> > > * Also when I click outside the checkbox it gets unchecked
> > >   image
> > > ```
> > 
> > 
> > So apparently the problem for you was not updating the browser after uploading new build version on the server. For me the dialog looks like this.
> > Screenshot_2019-12-04 Courses
> > Regarding the clicking outside I agree with @krusche that it is not a big issue and would take some time to fix. Especially, that we had this behavior all the time
> 
> Not eon the matter - I do not experience that issue
I gotta be very lucky or better said unlucky to be able to reproduce this ^^' Only happens when you click under the word "can" of the delete programming exercise popup as seen in the screenshot. Believe me, I refreshed the page multiple times this time XD I think last time I updated my browser way too early, that's why I couldn't see the changes.
